beacon-chain/sync/pending_blocks_queue.go

```diff
@@ -279,28 +301,76 @@ func (s *Service) deleteBlockFromPendingQueue(slot uint64, b *ethpb.SignedBeacon
 		newBlks = append(newBlks, blk)
 	}
 	if len(newBlks) == 0 {
-		delete(s.slotToPendingBlocks, slot)
-		return
+		s.slotToPendingBlocks.Delete(slotToCacheKey(slot))
+		return nil
+	}
+
+	// Decrease exp itme in proportion to how many blocks are still in the cache for slot key.
+	d := gcache.DefaultExpiration.Seconds() / float64(len(newBlks))
+	if err := s.slotToPendingBlocks.Replace(slotToCacheKey(slot), newBlks, time.Duration(d)); err != nil {
+		return err
 	}
-	s.slotToPendingBlocks[slot] = newBlks
 	delete(s.seenPendingBlocks, r)
+	return nil
 }
 
 // Insert block to the list in the pending queue using the slot as key.
 // Note: this helper is not thread safe.
-func (s *Service) insertBlockToPendingQueue(slot uint64, b *ethpb.SignedBeaconBlock, r [32]byte) {
+func (s *Service) insertBlockToPendingQueue(slot uint64, b *ethpb.SignedBeaconBlock, r [32]byte) error {
 	mutexasserts.AssertRWMutexLocked(&s.pendingQueueLock)
 
 	if s.seenPendingBlocks[r] {
-		return
+		return nil
 	}
 
-	_, ok := s.slotToPendingBlocks[slot]
-	if ok {
-		blks := s.slotToPendingBlocks[slot]
-		s.slotToPendingBlocks[slot] = append(blks, b)
-	} else {
-		s.slotToPendingBlocks[slot] = []*ethpb.SignedBeaconBlock{b}
+	if err := s.addPendingBlockToCache(b); err != nil {
+		return err
 	}
+
 	s.seenPendingBlocks[r] = true
+	return nil
+}
+
+// This returns signed beacon blocks given input key from slotToPendingBlocks.
+func (s *Service) pendingBlocksInCache(slot uint64) []*ethpb.SignedBeaconBlock {
+	k := slotToCacheKey(slot)
+	value, ok := s.slotToPendingBlocks.Get(k)
+	if !ok {
+		return []*ethpb.SignedBeaconBlock{}
+	}
+	blks, ok := value.([]*ethpb.SignedBeaconBlock)
+	if !ok {
+		return []*ethpb.SignedBeaconBlock{}
+	}
+	return blks
+}
+
+// This adds input signed beacon block to slotToPendingBlocks cache.
+func (s *Service) addPendingBlockToCache(b *ethpb.SignedBeaconBlock) error {
+	if b == nil || b.Block == nil {
+		return errors.New("nil block")
+	}
+
+	blks := s.pendingBlocksInCache(b.Block.Slot)
+
+	if len(blks) >= maxBlocksPerSlot {
+		return fmt.Errorf("could not add pending block to queue for slot %d", b.Block.Slot)
```

Comment:
wouldn't it be better to simply just return here ? without an error

Reply:
sounds good 👍